core/src/main/java/org/elasticsearch/rest/RestRequest.java

```diff
@@ -100,9 +140,53 @@ public final String path() {
 
     public abstract BytesReference content();
 
-    public abstract String header(String name);
+    /**
+     * Get the value of the header or {@code null} if not found. This method only retrieves the first header value if multiple values are
+     * sent. Use of {@link #getAllHeaderValues(String)} should be preferred
+     */
+    public final String header(String name) {
+        List<String> values = headers.get(name);
+        if (values != null && values.isEmpty() == false) {
+            return values.get(0);
+        }
+        return null;
+    }
+
+    /**
+     * Get all values for the header or {@code null} if the header was not found
+     */
+    public final List<String> getAllHeaderValues(String name) {
+        List<String> values = headers.get(name);
+        if (values != null) {
+            return Collections.unmodifiableList(values);
+        }
+        return null;
+    }
+
+    /**
+     * Get all of the headers and values associated with the headers. Modifications of this map are not supported.
+     */
+    public final Map<String, List<String>> getHeaders() {
+        return headers;
+    }
 
-    public abstract Iterable<Map.Entry<String, String>> headers();
+    /**
+     * The {@link XContentType} that was parsed from the {@code Content-Type} header. This value will be {@code null} in the case of
+     * a request without a valid {@code Content-Type} header, a request without content ({@link #hasContent()}, or a plain text request
+     */
+    @Nullable
+    public final XContentType getXContentType() {
+        return xContentType.get();
+    }
+
+    /**
+     * Sets the {@link XContentType}
+     * @deprecated this is only used to allow BWC with content-type detection
+     */
+    @Deprecated
+    final void setxContentType(XContentType xContentType) {
```

Comment:
This is needed for handling the BWC layer in the RestController. I think it makes sense to do it there but then we need a way to set the XContentType on the RestRequest. This will all be cleaner once bwc is removed.